File: ios/VineTrack/App/PinAisleGeometry.swift
import Foundation
import CoreLocation

/// Pure geometry for the core pin-location contract: which aisle (inter-row
/// driving path) an operator occupies, and which of that aisle's two physically
/// adjacent vine rows lies on their Left vs Right.
///
/// Everything here is derived from mapped row geometry and a validated heading.
/// Row-number ordering is never assumed — Left is not "the lower number", it is
/// whichever adjacent row actually sits to the operator's left.
///
/// Pure functions only. No store, no @MainActor, no I/O.
nonisolated enum PinAisleGeometry {

    /// The two physically adjacent rows bounding the operator's aisle.
    struct Aisle: Sendable, Equatable {
        /// Driving path identifier, the mean of the two adjacent row numbers
        /// (rows 32 and 33 -> 32.5). Derived from geometry, never assumed.
        let aisleNumber: Double
        /// Row whose centreline is nearest the fix.
        let nearRowNumber: Int
        /// Row on the far side of the fix, bounding the same aisle.
        let farRowNumber: Int
    }

    /// A row selected for a given operator side, with its snap onto that row's
    /// own centreline (never the aisle midline).
    struct RowSelection: Sendable {
        let rowNumber: Int
        let snapped: CLLocationCoordinate2D
        let distanceAlongMetres: Double
    }

    /// Minimum offset from a row centreline before an aisle can be claimed.
    /// A fix sitting essentially on the vine row itself does not establish
    /// which of the two neighbouring aisles the operator occupied.
    static let minimumAisleOffsetMetres: Double = 0.15

    /// Fallback aisle width ceiling when the block records no row spacing.
    static let fallbackMaxAisleWidthMetres: Double = 12.0

    /// Oldest compass sample still accepted as the operator's facing at the
    /// moment of capture. Heading freshness only — this never changes the GPS
    /// acceptance thresholds enforced by `LocationService`.
    static let maximumHeadingAgeSeconds: Double = 5.0

    /// True when `heading` is a usable recorded facing. A genuine 0° (North) is
    /// valid; nil, non-finite and out-of-range values are not turned into North.
    static func validHeading(_ heading: Double?) -> Double? {
        guard let heading, heading.isFinite, heading >= -0.0001, heading <= 360.0001 else {
            return nil
        }
        return normalizedDegrees(heading)
    }

    /// Validated facing with temporal evidence. A compass sample older than
    /// `maximumHeadingAgeSeconds` describes an earlier moment, not this capture,
    /// so it is discarded instead of frozen into the pin. A nil age means no age
    /// was reported and is not treated as evidence of staleness.
    static func validHeading(_ heading: Double?, ageSeconds: Double?) -> Double? {
        if let ageSeconds {
            guard ageSeconds.isFinite,
                  ageSeconds >= -0.5,
                  ageSeconds <= maximumHeadingAgeSeconds
            else { return nil }
        }
        return validHeading(heading)
    }

    /// True when the accepted fix is precise enough for mapped-aisle matching.
    ///
    /// Requiring the entire accuracy circle to fit on both sides of a ~3 m aisle
    /// made ordinary field GPS incapable of attaching almost every standalone
    /// pin. The mapped adjacent rows, containment and headland checks already
    /// constrain the candidate corridor, so accuracy is compared with the full
    /// mapped aisle width. An uncertainty spanning the aisle (or no valid
    /// accuracy at all) remains ambiguous and requires operator confirmation.
    static func uncertaintyFitsBetweenRows(
        horizontalAccuracyMetres: Double?,
        distanceToNearRowMetres: Double,
        distanceToFarRowMetres: Double
    ) -> Bool {
        guard let accuracy = horizontalAccuracyMetres,
              accuracy.isFinite,
              accuracy >= 0,
              distanceToNearRowMetres.isFinite,
              distanceToNearRowMetres > 0,
              distanceToFarRowMetres.isFinite,
              distanceToFarRowMetres > 0
        else { return false }
        return accuracy < distanceToNearRowMetres + distanceToFarRowMetres
    }

    /// Resolve the aisle physically containing `coordinate`: the nearest mapped
    /// row plus the nearest row on the opposite side of the fix, provided they
    /// are close enough to be a real aisle AND the reported GPS uncertainty is
    /// small enough to tell this aisle apart from its neighbours.
    ///
    /// Returns nil when the block has no mapped rows, the fix sits on a row
    /// centreline, the fix lies beyond the ends of the rows (a clamped endpoint
    /// projection proves nothing about containment), no plausible neighbour
    /// exists, or the uncertainty spans more than the aisle. Callers keep an
    /// honest unconfirmed attachment in every one of those cases.
    ///
    /// - Parameter horizontalAccuracyMetres: the fix's own reported accuracy
    ///   radius. Required evidence — nil or invalid resolves to no aisle.
    static func aisle(
        containing coordinate: CLLocationCoordinate2D,
        in paddock: Paddock,
        horizontalAccuracyMetres: Double?
    ) -> Aisle? {
        resolveAisle(
            containing: coordinate,
            in: paddock,
            horizontalAccuracyMetres: horizontalAccuracyMetres,
            requiresQualifiedAccuracy: true
        )
    }

    /// Browsing-only aisle estimate. It uses the same mapped-row containment,
    /// headland and width checks as capture, but does not claim capture-grade GPS
    /// certainty. Callers must label the result as approximate and never persist it.
    static func approximateAisle(
        containing coordinate: CLLocationCoordinate2D,
        in paddock: Paddock
    ) -> Aisle? {
        resolveAisle(
            containing: coordinate,
            in: paddock,
            horizontalAccuracyMetres: nil,
            requiresQualifiedAccuracy: false
        )
    }

    private static func resolveAisle(
        containing coordinate: CLLocationCoordinate2D,
        in paddock: Paddock,
        horizontalAccuracyMetres: Double?,
        requiresQualifiedAccuracy: Bool
    ) -> Aisle? {
        let rows = paddock.rows
        guard rows.count >= 2 else { return nil }

        let frame = MetricFrame(around: coordinate)
        let point = frame.project(coordinate)

        // Nearest row and its closest point to the fix.
        var nearest: (row: PaddockRow, closest: Projection, distance: Double)?
        for row in rows {
            let a = frame.project(row.startPoint.coordinate)
            let b = frame.project(row.endPoint.coordinate)
            guard let closest = closestPoint(on: a, b, to: point) else { continue }
            let distance = closest.point.distance(to: point)
            if nearest == nil || distance < (nearest?.distance ?? .greatestFiniteMagnitude) {
                nearest = (row, closest, distance)
            }
        }
        guard let near = nearest, near.distance >= minimumAisleOffsetMetres else { return nil }
        // Past the end of a row the nearest point is only its endpoint. That
        // clamped projection is a geometry artefact, never proof the operator
        // was between the rows, so a headland fix stays unconfirmed.
        guard !near.closest.clampedToEnd else { return nil }

        // Axis: the outward direction from the nearest row towards the fix.
        let axis = Point(x: point.x - near.closest.point.x, y: point.y - near.closest.point.y)
            .normalized()
        guard let axis else { return nil }
        let fixOffset = near.distance

        // The aisle's far side is the closest row lying beyond the fix along
        // that same axis. Angled rows are handled because each candidate is
        // measured at its own closest point to the fix.
        var far: (row: PaddockRow, closest: Projection, offset: Double)?
        for row in rows where row.number != near.row.number {
            let a = frame.project(row.startPoint.coordinate)
            let b = frame.project(row.endPoint.coordinate)
            guard let closest = closestPoint(on: a, b, to: point), !closest.clampedToEnd else { continue }
            let offset = (closest.point.x - near.closest.point.x) * axis.x
                + (closest.point.y - near.closest.point.y) * axis.y
            guard offset > fixOffset else { continue }
            if far == nil || offset < (far?.offset ?? .greatestFiniteMagnitude) {
                far = (row, closest, offset)
            }
        }
        guard let far else { return nil }

        // Reject implausibly wide pairs: a missing neighbour must never be
        // invented by pairing two rows that don't actually form one aisle.
        let maxWidth = paddock.rowWidth > 0 ? paddock.rowWidth * 2.5 : fallbackMaxAisleWidthMetres
        guard far.offset <= maxWidth else { return nil }

        // Map matching has already established the containing adjacent-row
        // corridor. Reject only when reported uncertainty spans that full aisle;
        // tighter single-fix evidence may attach without requiring trip state.
        let farDistance = far.closest.point.distance(to: point)
        if requiresQualifiedAccuracy {
            guard uncertaintyFitsBetweenRows(
                horizontalAccuracyMetres: horizontalAccuracyMetres,
                distanceToNearRowMetres: near.distance,
                distanceToFarRowMetres: farDistance
            ) else { return nil }
        }

        return Aisle(
            aisleNumber: (Double(near.row.number) + Double(far.row.number)) / 2.0,
            nearRowNumber: near.row.number,
            farRowNumber: far.row.number
        )
    }

    /// Choose which of two candidate rows lies on the operator's `side` given a
    /// validated `heading`, then snap the fix onto that row's own centreline.
    ///
    /// Returns nil when the heading is unusable or the candidate geometry is
    /// missing — an unconfirmed side is never guessed.
    static func rowOnSide(
        rowNumbers: (Int, Int),
        coordinate: CLLocationCoordinate2D,
        heading: Double?,
        operatorSide: PinSide,
        in paddock: Paddock,
        useAisleMidpointReference: Bool = false
    ) -> RowSelection? {
        guard let heading = validHeading(heading) else { return nil }
        guard let first = paddock.rows.first(where: { $0.number == rowNumbers.0 }),
              let second = paddock.rows.first(where: { $0.number == rowNumbers.1 }),
              first.number != second.number
        else { return nil }

        let frame = MetricFrame(around: coordinate)
        let point = frame.project(coordinate)
        guard let firstClosest = closestPoint(
            on: frame.project(first.startPoint.coordinate),
            frame.project(first.endPoint.coordinate),
            to: point
        )?.point, let secondClosest = closestPoint(
            on: frame.project(second.startPoint.coordinate),
            frame.project(second.endPoint.coordinate),
            to: point
        )?.point else { return nil }

        // A locked aisle survives a brief lateral GPS outlier. Use the mapped
        // aisle midpoint only to determine physical Left/Right; the raw current
        // fix remains unchanged and is still projected onto the selected row.
        let sideReference = useAisleMidpointReference
            ? Point(x: (firstClosest.x + secondClosest.x) / 2, y: (firstClosest.y + secondClosest.y) / 2)
            : point
        let leftBearing = normalizedDegrees(heading - 90)
        let firstBearing = bearing(from: sideReference, to: firstClosest)
        let secondBearing = bearing(from: sideReference, to: secondClosest)
        guard let firstBearing, let secondBearing else { return nil }

        let firstIsLeft = abs(signedAngularDifference(firstBearing, leftBearing)) < 90
        let secondIsLeft = abs(signedAngularDifference(secondBearing, leftBearing)) < 90
        // The two rows must genuinely lie on opposite sides of the operator.
        guard firstIsLeft != secondIsLeft else { return nil }

        let chosen: PaddockRow = {
            switch operatorSide {
            case .left: return firstIsLeft ? first : second
            case .right: return firstIsLeft ? second : first
            }
        }()

        guard let snap = RowGuidance.snapToRow(
            coordinate: coordinate,
            rowNumber: chosen.number,
            in: paddock
        ) else { return nil }

        return RowSelection(
            rowNumber: chosen.number,
            snapped: snap.snapped,
            distanceAlongMetres: snap.distanceAlongMetres
        )
    }

    /// The two vine rows bounding an explicit driving path (e.g. 32.5 -> 32 and
    /// 33). Returns nil unless both rows exist in the block's mapped geometry.
    static func rowsBounding(path: Double, in paddock: Paddock) -> (Int, Int)? {
        let lower = Int(floor(path))
        let upper = Int(ceil(path))
        guard lower != upper else { return nil }
        guard paddock.rows.contains(where: { $0.number == lower }),
              paddock.rows.contains(where: { $0.number == upper })
        else { return nil }
        return (lower, upper)
    }

    // MARK: - Local metric frame

    private struct Point {
        /// Metres east of the frame origin.
        let x: Double
        /// Metres north of the frame origin.
        let y: Double

        func distance(to other: Point) -> Double {
            let dx = x - other.x
            let dy = y - other.y
            return (dx * dx + dy * dy).squareRoot()
        }

        func normalized() -> Point? {
            let length = (x * x + y * y).squareRoot()
            guard length > 1e-9 else { return nil }
            return Point(x: x / length, y: y / length)
        }
    }

    private struct MetricFrame {
        let originLat: Double
        let originLon: Double
        let metresPerDegreeLat: Double = 111_320.0
        let metresPerDegreeLon: Double

        init(around coordinate: CLLocationCoordinate2D) {
            originLat = coordinate.latitude
            originLon = coordinate.longitude
            metresPerDegreeLon = 111_320.0 * cos(coordinate.latitude * .pi / 180.0)
        }

        func project(_ coordinate: CLLocationCoordinate2D) -> Point {
            Point(
                x: (coordinate.longitude - originLon) * metresPerDegreeLon,
                y: (coordinate.latitude - originLat) * metresPerDegreeLat
            )
        }
    }

    /// Closest point on a segment, plus whether the projection had to be
    /// clamped to an endpoint — i.e. the fix lies beyond that end of the row.
    private struct Projection {
        let point: Point
        let clampedToEnd: Bool
    }

    private static func closestPoint(on a: Point, _ b: Point, to p: Point) -> Projection? {
        let dx = b.x - a.x
        let dy = b.y - a.y
        let lengthSquared = dx * dx + dy * dy
        guard lengthSquared > 1e-9 else { return nil }
        let rawT = ((p.x - a.x) * dx + (p.y - a.y) * dy) / lengthSquared
        let t = max(0, min(1, rawT))
        let tolerance = 1e-9
        return Projection(
            point: Point(x: a.x + t * dx, y: a.y + t * dy),
            clampedToEnd: rawT < -tolerance || rawT > 1 + tolerance
        )
    }

    /// True bearing (0–360°) of the vector `from` -> `to` in the local frame.
    private static func bearing(from: Point, to: Point) -> Double? {
        let east = to.x - from.x
        let north = to.y - from.y
        guard (east * east + north * north).squareRoot() > 1e-6 else { return nil }
        return normalizedDegrees(atan2(east, north) * 180.0 / .pi)
    }

    /// Signed difference (a − b) wrapped to (−180, 180].
    static func signedAngularDifference(_ a: Double, _ b: Double) -> Double {
        var diff = (a - b).truncatingRemainder(dividingBy: 360)
        if diff > 180 { diff -= 360 }
        if diff <= -180 { diff += 360 }
        return diff
    }

    static func normalizedDegrees(_ value: Double) -> Double {
        let v = value.truncatingRemainder(dividingBy: 360)
        return v < 0 ? v + 360 : v
    }
}

File: android-vinetrack/app/src/main/java/com/rork/vinetrack/data/PinAisleGeometry.kt
package com.rork.vinetrack.data

import com.rork.vinetrack.data.model.Paddock
import kotlin.math.abs
import kotlin.math.atan2
import kotlin.math.cos
import kotlin.math.sqrt

/**
 * Pure geometry for the core pin-location contract (Android mirror of the iOS
 * `PinAisleGeometry`): which aisle (inter-row driving path) an operator
 * occupies, and which of that aisle's two physically adjacent vine rows lies on
 * their Left vs Right.
 *
 * Everything is derived from mapped row geometry and a validated heading. Row
 * numbering order is never assumed — Left is not "the lower number", it is
 * whichever adjacent row actually sits to the operator's left.
 *
 * See docs/core-pin-location-contract.md.
 */
object PinAisleGeometry {

    private const val METRES_PER_DEG_LAT = 111_320.0

    /**
     * Minimum offset from a row centreline before an aisle can be claimed. A fix
     * sitting essentially on the vine row itself does not establish which of the
     * two neighbouring aisles the operator occupied.
     */
    const val MIN_AISLE_OFFSET_M: Double = 0.15

    /** Aisle-width ceiling used when the block records no row spacing. */
    const val FALLBACK_MAX_AISLE_WIDTH_M: Double = 12.0

    /**
     * Oldest compass sample still accepted as the operator's facing at capture.
     * Heading freshness only — GPS acceptance thresholds are unchanged.
     */
    const val MAX_HEADING_AGE_MS: Long = 5_000L

    /**
     * Slowest travel that still proves a direction of travel. Below this the
     * GPS course is noise from a stationary or crawling machine and must never
     * be labelled as the operator's facing.
     */
    const val MIN_COURSE_SPEED_MPS: Double = 1.0

    /** The two physically adjacent rows bounding the operator's aisle. */
    data class Aisle(
        /** Driving path identifier: mean of the two adjacent row numbers (32 + 33 -> 32.5). */
        val aisleNumber: Double,
        /** Row whose centreline is nearest the fix. */
        val nearRowNumber: Int,
        /** Row on the far side of the fix bounding the same aisle. */
        val farRowNumber: Int,
    )

    /** A row selected for an operator side, snapped onto that row's own centreline. */
    data class RowSelection(
        val rowNumber: Int,
        val snappedLatitude: Double,
        val snappedLongitude: Double,
        val alongRowDistanceM: Double,
    )

    /**
     * A usable recorded facing, or null. A genuine 0.0 (North) is valid; missing,
     * non-finite and out-of-range values are never turned into North.
     */
    fun validHeading(headingDegrees: Double?): Double? {
        val h = headingDegrees ?: return null
        if (!h.isFinite() || h < -0.0001 || h > 360.0001) return null
        return normalizedDegrees(h)
    }

    /**
     * Validated facing with temporal evidence. A compass sample older than
     * [MAX_HEADING_AGE_MS] describes an earlier moment, not this capture, so it
     * is discarded instead of frozen into the pin. A null age means no age was
     * reported and is not treated as evidence of staleness.
     */
    fun validHeading(headingDegrees: Double?, ageMs: Long?): Double? {
        if (ageMs != null && (ageMs < -500L || ageMs > MAX_HEADING_AGE_MS)) return null
        return validHeading(headingDegrees)
    }

    /**
     * A GPS travel course qualified as the operator's facing. Numeric validity
     * alone is not enough: a stationary or reversing machine reports a course
     * that is not where the operator is looking, so a course is accepted only
     * with genuine forward travel evidence.
     */
    fun qualifiedCourseHeading(
        bearingDegrees: Double?,
        speedMetresPerSecond: Double?,
        hasForwardTravelEvidence: Boolean,
    ): Double? {
        if (!hasForwardTravelEvidence) return null
        val speed = speedMetresPerSecond ?: return null
        if (!speed.isFinite() || speed < MIN_COURSE_SPEED_MPS) return null
        return validHeading(bearingDegrees)
    }

    /**
     * Freeze one facing for both row selection and persistence. Compass evidence
     * must carry a current sensor timestamp. GPS course is accepted only when a
     * caller has independent evidence that travel is forward, not reversing.
     */
    fun frozenCaptureHeading(
        compassHeadingDegrees: Double?,
        compassObservedAtElapsedRealtimeNanos: Long?,
        captureElapsedRealtimeNanos: Long,
        courseDegrees: Double?,
        speedMetresPerSecond: Double?,
        hasForwardTravelEvidence: Boolean,
    ): Double? {
        val compassAgeMs = compassObservedAtElapsedRealtimeNanos?.let { observedAt ->
            (captureElapsedRealtimeNanos - observedAt) / 1_000_000L
        }
        val compass = if (compassObservedAtElapsedRealtimeNanos == null) {
            null
        } else {
            validHeading(compassHeadingDegrees, compassAgeMs)
        }
        return compass ?: qualifiedCourseHeading(
            courseDegrees,
            speedMetresPerSecond,
            hasForwardTravelEvidence,
        )
    }

    /**
     * True when an accepted fix is precise enough for mapped-aisle matching.
     * Requiring the whole accuracy circle to fit on both sides of a ~3 m aisle
     * rejected ordinary field GPS. Adjacent-row containment and headland checks
     * already constrain the corridor, so accuracy is compared with its full
     * mapped width. Missing accuracy or uncertainty spanning the aisle remains
     * ambiguous and requires confirmation.
     */
    fun uncertaintyFitsBetweenRows(
        accuracyMetres: Double?,
        distanceToNearRowMetres: Double,
        distanceToFarRowMetres: Double,
    ): Boolean {
        val accuracy = accuracyMetres ?: return false
        if (!accuracy.isFinite() || accuracy < 0.0) return false
        if (!distanceToNearRowMetres.isFinite() || distanceToNearRowMetres <= 0.0) return false
        if (!distanceToFarRowMetres.isFinite() || distanceToFarRowMetres <= 0.0) return false
        return accuracy < distanceToNearRowMetres + distanceToFarRowMetres
    }

    /**
     * Resolve the aisle physically containing the fix: the nearest mapped row
     * plus the nearest row lying beyond the fix on the same axis, provided the
     * pair is close enough to be a real aisle.
     *
     * Returns null when the block has no mapped row geometry, the fix sits on a
     * row centreline, the fix lies beyond the ends of the rows (a clamped
     * endpoint projection proves nothing about containment), no plausible
     * neighbour exists, or the reported uncertainty spans more than the aisle.
     *
     * [accuracyMetres] is required evidence — null or invalid resolves to no
     * aisle, and the caller keeps an honest unconfirmed attachment.
     */
    fun aisleContaining(
        paddock: Paddock?,
        latitude: Double,
        longitude: Double,
        accuracyMetres: Double?,
    ): Aisle? = resolveAisle(
        paddock = paddock,
        latitude = latitude,
        longitude = longitude,
        accuracyMetres = accuracyMetres,
        requiresQualifiedAccuracy = true,
    )

    /**
     * Browsing-only aisle estimate. It retains mapped-row containment, headland
     * and width checks but is never capture evidence and must never be persisted.
     */
    fun approximateAisle(
        paddock: Paddock?,
        latitude: Double,
        longitude: Double,
    ): Aisle? = resolveAisle(
        paddock = paddock,
        latitude = latitude,
        longitude = longitude,
        accuracyMetres = null,
        requiresQualifiedAccuracy = false,
    )

    private fun resolveAisle(
        paddock: Paddock?,
        latitude: Double,
        longitude: Double,
        accuracyMetres: Double?,
        requiresQualifiedAccuracy: Boolean,
    ): Aisle? {
        val rows = paddock?.rows
            ?.filter { it.startPoint != null && it.endPoint != null }
            ?.takeIf { it.size >= 2 }
            ?: return null

        val frame = MetricFrame(latitude, longitude)
        val point = frame.project(latitude, longitude)

        var nearRow: Int? = null
        var nearClosest: Point? = null
        var nearDistance = Double.MAX_VALUE
        for (row in rows) {
            val projection = projectOntoRow(frame, row, point) ?: continue
            val distance = projection.point.distanceTo(point)
            if (distance < nearDistance) {
                nearDistance = distance
                nearRow = row.number
                nearClosest = if (projection.clampedToEnd) null else projection.point
            }
        }
        val nearNumber = nearRow ?: return null
        // Past the end of a row the nearest point is only its endpoint. That
        // clamped projection is a geometry artefact, never proof the operator
        // was between the rows, so a headland fix stays unconfirmed.
        val nearPoint = nearClosest ?: return null
        if (nearDistance < MIN_AISLE_OFFSET_M) return null

        // Axis: outward direction from the nearest row towards the fix.
        val axis = Point(point.x - nearPoint.x, point.y - nearPoint.y).normalized() ?: return null

        var farRow: Int? = null
        var farOffset = Double.MAX_VALUE
        var farDistance = Double.MAX_VALUE
        for (row in rows) {
            if (row.number == nearNumber) continue
            val projection = projectOntoRow(frame, row, point) ?: continue
            if (projection.clampedToEnd) continue
            val closest = projection.point
            val offset = (closest.x - nearPoint.x) * axis.x + (closest.y - nearPoint.y) * axis.y
            // Must lie beyond the fix, i.e. on the opposite side of the aisle.
            if (offset <= nearDistance) continue
            if (offset < farOffset) {
                farOffset = offset
                farDistance = closest.distanceTo(point)
                farRow = row.number
            }
        }
        val farNumber = farRow ?: return null

        val rowWidth = paddock.rowWidth?.takeIf { it > 0 }
        val maxWidth = if (rowWidth != null) rowWidth * 2.5 else FALLBACK_MAX_AISLE_WIDTH_M
        if (farOffset > maxWidth) return null

        // Map matching already established the containing adjacent-row corridor.
        // Reject only when uncertainty spans that full mapped aisle.
        if (requiresQualifiedAccuracy && !uncertaintyFitsBetweenRows(accuracyMetres, nearDistance, farDistance)) return null

        return Aisle(
            aisleNumber = (nearNumber.toDouble() + farNumber.toDouble()) / 2.0,
            nearRowNumber = nearNumber,
            farRowNumber = farNumber,
        )
    }

    /**
     * Choose which of two candidate rows lies on the operator's [side] for a
     * validated [headingDegrees], then snap the fix onto that row's own
     * centreline (never the aisle midline).
     *
     * Returns null when the heading is unusable, the rows are missing, or the two
     * candidates do not actually lie on opposite sides of the operator — an
     * unconfirmed side is never guessed.
     */
    fun rowOnSide(
        paddock: Paddock?,
        rowNumbers: Pair<Int, Int>,
        latitude: Double,
        longitude: Double,
        headingDegrees: Double?,
        side: String?,
        useAisleMidpointReference: Boolean = false,
    ): RowSelection? {
        val heading = validHeading(headingDegrees) ?: return null
        val cleanSide = side?.trim()?.lowercase()?.takeIf { it == "left" || it == "right" } ?: return null
        if (rowNumbers.first == rowNumbers.second) return null
        val rows = paddock?.rows?.filter { it.startPoint != null && it.endPoint != null } ?: return null
        val first = rows.firstOrNull { it.number == rowNumbers.first } ?: return null
        val second = rows.firstOrNull { it.number == rowNumbers.second } ?: return null

        val frame = MetricFrame(latitude, longitude)
        val point = frame.project(latitude, longitude)
        val firstClosest = closestPointOnRow(frame, first, point) ?: return null
        val secondClosest = closestPointOnRow(frame, second, point) ?: return null

        // A locked aisle survives a brief lateral GPS outlier. The mapped
        // midpoint determines only Left/Right; snapping still projects the
        // unchanged current fix onto the selected row.
        val sideReference = if (useAisleMidpointReference) {
            Point((firstClosest.x + secondClosest.x) / 2.0, (firstClosest.y + secondClosest.y) / 2.0)
        } else point
        val leftBearing = normalizedDegrees(heading - 90.0)
        val firstBearing = bearing(sideReference, firstClosest) ?: return null
        val secondBearing = bearing(sideReference, secondClosest) ?: return null
        val firstIsLeft = abs(signedAngularDifference(firstBearing, leftBearing)) < 90.0
        val secondIsLeft = abs(signedAngularDifference(secondBearing, leftBearing)) < 90.0
        // The two rows must genuinely lie on opposite sides of the operator.
        if (firstIsLeft == secondIsLeft) return null

        val chosen = when (cleanSide) {
            "left" -> if (firstIsLeft) first else second
            else -> if (firstIsLeft) second else first
        }
        val chosenStart = chosen.startPoint ?: return null
        val chosenEnd = chosen.endPoint ?: return null
        val a = frame.project(chosenStart.latitude, chosenStart.longitude)
        val b = frame.project(chosenEnd.latitude, chosenEnd.longitude)
        val snap = snapOnto(a, b, point) ?: return null
        val snapped = frame.unproject(snap.point)
        return RowSelection(
            rowNumber = chosen.number,
            snappedLatitude = snapped.first,
            snappedLongitude = snapped.second,
            alongRowDistanceM = snap.distanceAlong,
        )
    }

    /**
     * The two vine rows bounding an explicit driving path (32.5 -> 32 and 33).
     * Null unless both rows exist in the block's mapped geometry.
     */
    fun rowsBoundingPath(paddock: Paddock?, path: Double): Pair<Int, Int>? {
        val lower = kotlin.math.floor(path).toInt()
        val upper = kotlin.math.ceil(path).toInt()
        if (lower == upper) return null
        val rows = paddock?.rows?.filter { it.startPoint != null && it.endPoint != null } ?: return null
        if (rows.none { it.number == lower } || rows.none { it.number == upper }) return null
        return lower to upper
    }

    // MARK: - Local metric frame

    private data class Point(val x: Double, val y: Double) {
        fun distanceTo(other: Point): Double {
            val dx = x - other.x
            val dy = y - other.y
            return sqrt(dx * dx + dy * dy)
        }

        fun normalized(): Point? {
            val length = sqrt(x * x + y * y)
            if (length <= 1e-9) return null
            return Point(x / length, y / length)
        }
    }

    private class MetricFrame(private val originLat: Double, private val originLon: Double) {
        private val metresPerDegLon = METRES_PER_DEG_LAT * cos(originLat * Math.PI / 180.0)

        fun project(lat: Double, lon: Double): Point = Point(
            x = (lon - originLon) * metresPerDegLon,
            y = (lat - originLat) * METRES_PER_DEG_LAT,
        )

        fun unproject(point: Point): Pair<Double, Double> = Pair(
            originLat + point.y / METRES_PER_DEG_LAT,
            originLon + point.x / metresPerDegLon,
        )
    }

    private data class Snap(
        val point: Point,
        val distanceAlong: Double,
        /** True when the fix lies beyond an end of the segment. */
        val clampedToEnd: Boolean,
    )

    private fun closestPointOnRow(
        frame: MetricFrame,
        row: com.rork.vinetrack.data.model.PaddockRow,
        point: Point,
    ): Point? = projectOntoRow(frame, row, point)?.point

    private fun projectOntoRow(
        frame: MetricFrame,
        row: com.rork.vinetrack.data.model.PaddockRow,
        point: Point,
    ): Snap? {
        val s = row.startPoint ?: return null
        val e = row.endPoint ?: return null
        val a = frame.project(s.latitude, s.longitude)
        val b = frame.project(e.latitude, e.longitude)
        return snapOnto(a, b, point)
    }

    private fun snapOnto(a: Point, b: Point, p: Point): Snap? {
        val dx = b.x - a.x
        val dy = b.y - a.y
        val lengthSquared = dx * dx + dy * dy
        if (lengthSquared <= 1e-9) return null
        val length = sqrt(lengthSquared)
        val rawT = ((p.x - a.x) * dx + (p.y - a.y) * dy) / lengthSquared
        val t = rawT.coerceIn(0.0, 1.0)
        val tolerance = 1e-9
        return Snap(
            point = Point(a.x + t * dx, a.y + t * dy),
            distanceAlong = t * length,
            clampedToEnd = rawT < -tolerance || rawT > 1.0 + tolerance,
        )
    }

    /** True bearing (0–360) of the vector from -> to in the local frame. */
    private fun bearing(from: Point, to: Point): Double? {
        val east = to.x - from.x
        val north = to.y - from.y
        if (sqrt(east * east + north * north) <= 1e-6) return null
        return normalizedDegrees(Math.toDegrees(atan2(east, north)))
    }

    /** Signed difference (a - b) wrapped to (-180, 180]. */
    fun signedAngularDifference(a: Double, b: Double): Double {
        var diff = (a - b) % 360.0
        if (diff > 180.0) diff -= 360.0
        if (diff <= -180.0) diff += 360.0
        return diff
    }

    fun normalizedDegrees(value: Double): Double {
        val v = value % 360.0
        return if (v < 0) v + 360.0 else v
    }
}
